graphbolt/src/cuda/index_select_impl.cu

```diff
@@ -48,6 +53,276 @@ std::pair<torch::Tensor, torch::Tensor> Sort(
+        // A single cache line has num_elements items, we add num_elements - 1
+        // to ensure there is enough slack to move forward or backward by
+        // num_elements - 1 items if the performed access is not aligned.
+        (indptr_t)(in_degree[perm ? perm[row % num_rows] : row] + num_elements - 1));
+  }
+};
+
+template <typename indptr_t, typename indices_t>
+__global__ void _CSRRowWiseOneHopExtractorAlignedKernel(
+    const indptr_t hop_size, const int64_t num_rows,
+    const indptr_t* const indptr, const indptr_t* const sub_indptr,
+    const indptr_t* const sub_indptr_aligned, const indices_t* const indices,
+    indices_t* const hop, const int64_t* const perm) {
+  indptr_t tx = static_cast<indptr_t>(blockIdx.x) * blockDim.x + threadIdx.x;
+  const int stride_x = gridDim.x * blockDim.x;
+
+  while (tx < hop_size) {
+    const auto rpos_ = cuda::UpperBound(sub_indptr_aligned, num_rows, tx) - 1;
+    const auto rpos = perm ? perm[rpos_] : rpos_;
+    const auto out_row = sub_indptr[rpos];
+    const auto d = sub_indptr[rpos + 1] - out_row;
+    const int offset =
+        ((size_t)(indices + indptr[rpos] - sub_indptr_aligned[rpos_]) %
+         GPU_CACHE_LINE_SIZE) /
+        sizeof(indices_t);
+    const auto rofs = tx - sub_indptr_aligned[rpos_] - offset;
+    if (rofs >= 0 && rofs < d) {
+      const auto in_idx = indptr[rpos] + rofs;
+      assert((size_t)(indices + in_idx - tx) % GPU_CACHE_LINE_SIZE == 0);
+      const auto u = indices[in_idx];
+      hop[out_row + rofs] = u;
+    }
+    tx += stride_x;
+  }
+}
+
+// Given rows and indptr, computes:
+// inrow_indptr[i] = indptr[rows[i]];
+// in_deg[i] = indptr[rows[i] + 1] - indptr[rows[i]];
+template <typename indptr_t, typename nodes_t>
+struct DegreeFunc {
+  const nodes_t* rows;
+  const indptr_t* indptr;
+  indptr_t* in_deg;
+  indptr_t* inrow_indptr;
+  __host__ __device__ auto operator()(int64_t tIdx) {
+    const auto out_row = rows[tIdx];
+    const auto indptr_val = indptr[out_row];
+    const auto degree = indptr[out_row + 1] - indptr_val;
+    in_deg[tIdx] = degree;
+    inrow_indptr[tIdx] = indptr_val;
+  }
+};
+
+struct PairSum {
+  template <typename indptr_t>
+  __host__ __device__ auto operator()(
+      thrust::tuple<indptr_t, indptr_t> a,
+      thrust::tuple<indptr_t, indptr_t> b) {
+    return thrust::make_tuple(
+        thrust::get<0>(a) + thrust::get<0>(b),
+        thrust::get<1>(a) + thrust::get<1>(b));
+  };
+};
+
+std::tuple<torch::Tensor, torch::Tensor> UVAIndexSelectCSCImpl(
+    torch::Tensor indptr, torch::Tensor indices, torch::Tensor nodes) {
+  // Sorting nodes so that accesses over PCI-e are more regular.
+  const auto [sorted, perm_tensor] =
+      Sort(nodes, cuda::NumberOfBits(indptr.size(0) - 1));
+  const auto perm = perm_tensor.data_ptr<int64_t>();
+
+  cuda::CUDAWorkspaceAllocator allocator;
+  auto stream = c10::cuda::getDefaultCUDAStream();
+  const auto exec_policy = thrust::cuda::par_nosync(allocator).on(stream);
```

Comment:
Please use a utility function to wrap the definition of the allocator, e.g., `auto allocator = cuda::BuildAllocator()`.